### PR TITLE
Sturdy up the API and add PDF417

### LINKED ISSUE #1: Add remaining zint generators (pdf417)
pdf417 is commonly used by banking applications so it would be very useful for generating invoices. But in general, given that there's already an enum of supported generators, why not add all of them.

I created a local copy of the plugin and added it, it ended up being a 1 line change for the pdf417 one.

### PR DESCRIPTION
**Pull request by Caellian**
I started adding PDF417 utility method, and then did a few other things. One day later:

- Allowed passing `Options` to Zint.
  - A lot of codes weren't constructible previously because they require additional configuration.
- Added utility functions for configuration.
  - This makes it easier and safer to use right `Options` values.
- Documented majority of barcode encoders in the manual.
- Replaced `anyhow` with `thiserror` to provide descriptive error messages.
  - Old ones were a single word that would be a pain to pin down in a lot of normal use cases.
- Removed some unused derives to speed up compilation.
  - e.g. `Options` don't need `Serialize` as they'll never be passed back to typst; Errors almost never require `Clone` and `Copy`, and adding them imposes restrictions that can't be satisfied in a lot of cases; ...
- Moved simple functions to typst and used composition to reduce code complexity.
- Marked `Symbology` as `#[non_exhaustive]` in bindings as it depends on Zint values.
  - Doesn't affect the current code, but ensures that future code will have to handle any unimplemented values in some way, in case Zint updates and the bindings don't (yet).
- Made `Options` fields public to allow testing and `Options` default constructible to simplify use from Rust code.
  - I faced some weird errors while trying to use Zint (my fault), so I tried writing tests and it was impossible to do so, so I fixed the API a little.
- Added a few constants to `lib.typ` to allow dependants to use named values instead of magic numbers.
  - I think these are mostly fine, but I'm not sure what `DM_ISO_144` and `ULTRA_COMPRESSION` do so I just left them in case they're needed.
- Fixed errors copying Options strings into C buffers.
  - Previously used `copy_from_slice` expected source and destination buffers to be of same size which caused the library to panic when any of the string options were set.
- Add more flexible Color deserialization logic.
  - The field can now deserialize form a hex string, or a byte array, or a dictionary.
- Turned `Option3` into an union so that it matches Zint better - the actual meaning of the value is determined by the `Options` object.
- Added more flexible Option3 deserialization - accepts either a valid int value (any of) or a textual representation.
- Made enum discriminants match the values they were representing.
  - This makes it very easy to do typst-rust-c conversion in any direction and makes things more consistent.
  - Used serde aliases to handle legacy names so this part is still backwards compatible.
- Use `hex` crate for faster Color conversion.
  - `hex` was needed for parsing/deserializing the colors, so I used it for string generation as well as it's faster and generates less code than std `format!` macro.
- Added a table with `Options` names and values to the manual so people can look it up there instead of having to look at the sources.

This closes #1.

## Future Work

- I left some more complicated barcodes commented out in the docs, they should be documented at some point, but it's a lot of work tracking down all the details and I'm short on time atm.
  - I made sure to include one of them anyway for reference.
- I left a few TODO comments in places where I didn't know what to do.

### CODE REVIEW

**Comment by Enter-tainer** [comment]
looks cool! i'm busy with work these days but i will take a look within a week. thank you for such a big improvement!

**Comment by Caellian** [comment]
You're welcome. Let me know if there's any change you don't like or want more info on when you get at it.

File: typst-package/lib.typ
Comment (Enter-tainer) [inline typst-package/lib.typ:1]: BTW you can add yourself to `authors` in `typst.toml`

File: typst-package/manual.typ [hunk 1]
```diff
@@ -1,435 +1,358 @@
 #import "./lib.typ": *
+#import "./lib.typ"
+#import "@preview/tidy:0.2.0"
+#import "@preview/tablex:0.0.8": tablex
+#import "./tidy_style.typ"
+
+// TODO: replace tidy_style with tidy.styles.default next version
+#set page(
```
Comment (Enter-tainer) [inline typst-package/manual.typ:8]: we can add a white background in page so texts are still visible when github is in dark theme

File: assets/doc-18.svg
Comment (Enter-tainer) [inline assets/doc-18.svg:1]: this page looks very similar to doc-17.svg. I guess the doc initially has 18 pages and reduced to 17 then. 🤣

File: assets/doc-01.svg
Comment (Enter-tainer) [inline assets/doc-01.svg:1]: Could you also edit the readme and add manual pages to it? I also wonder if we could switch to a pdf manual. SVG ones might make sense when it is only a few pages. But when it gets longer, I think a pdf manual would be better.

Both pdf or svg are OK to me.

**Review by Enter-tainer — APPROVED** [review]
LGTM, this is definitely a huge improvement to this packge!

File: assets/doc-18.svg
Comment (Caellian) [reply assets/doc-18.svg:1]: Right, didn't delete that page while.
I added a `rm` command to `./build.sh` to cleanup old assets in future.

File: typst-package/lib.typ
Comment (Caellian) [reply typst-package/lib.typ:1]: Thanks for the invite 😊. Didn't plan on making such a large PR 😆

File: assets/doc-01.svg
Comment (Caellian) [reply assets/doc-01.svg:1]: I like what you did with including SVGs better than just linking the manual, sadly GitHub doesn't allow a `README.pdf`. I added the white background.

Comment (Caellian) [reply assets/doc-01.svg:1]: Right... links don't work on SVGs. I ended up adding a preview which links to the manual.

**Comment by Enter-tainer** [comment]
thank you! i will publish a new version when i leave work

**Pull request merged**